packages/protocol/test/tokenomics/blockFee.test.ts

```diff
@@ -0,0 +1,135 @@
+        const { maxNumBlocks, commitConfirmations } = await taikoL1.getConfig();
+        // wait for one period of halving to occur, so fee is not 0.
+        const blockIdsToNumber: any = {};
+
+        // set up a proposer to continually propose new blocks
+        const proposer = new Proposer(
+            taikoL1.connect(proposerSigner),
+            l2Provider,
+            commitConfirmations.toNumber(),
+            maxNumBlocks.toNumber(),
+            0,
+            proposerSigner
+        );
+
+        // get the initiaal tkoBalance, which should decrease every block proposal
+        let lastProposerTkoBalance = await tkoTokenL1.balanceOf(
+            await proposerSigner.getAddress()
+        );
+
+        // do the same for the blockFee, which should increase every block proposal
+        // with proofs not being submitted.
+        // we want to wait for enough blocks until the blockFee is no longer 0, then run our
+        // tests.
+        let lastBlockFee = await taikoL1.getBlockFee();
+        while (lastBlockFee.eq(0)) {
+            await sleep(500);
+            lastBlockFee = await taikoL1.getBlockFee();
+        }
+
+        let lastProofReward = BigNumber.from(0);
+
+        let hasFailedAssertions: boolean = false;
+        let blocksProposed: number = 0;
+        // every time a l2 block is created, we should try to propose it on L1.
+        l2Provider.on("block", async (blockNumber: number) => {
+            if (blockNumber <= genesisHeight) return;
+            try {
+                const { newProposerTkoBalance, newBlockFee, newProofReward } =
+                    await onNewL2Block(
+                        l2Provider,
+                        blockNumber,
+                        proposer,
+                        blockIdsToNumber,
+                        taikoL1,
+                        proposerSigner,
+                        tkoTokenL1
+                    );
+
+                expect(
+                    newProposerTkoBalance.lt(lastProposerTkoBalance)
+                ).to.be.eq(true);
+                expect(newBlockFee.gt(lastBlockFee)).to.be.eq(true);
+                expect(newProofReward.gt(lastProofReward)).to.be.eq(true);
+
+                lastBlockFee = newBlockFee;
+                lastProofReward = newProofReward;
+                lastProposerTkoBalance = newProposerTkoBalance;
+                blocksProposed++;
+            } catch (e) {
+                hasFailedAssertions = true;
+                console.error(e);
+                throw e;
+            }
+        });
+
+        // wait until one roudn of blocks are proposed, then expect none of them to have
+        // failed their assertions.
+        while (blocksProposed < maxNumBlocks.toNumber() - 1) {
```

Comment:
not sure but if we use Node.js std lib's [`EventEmitter`](https://nodejs.org/api/events.html#class-eventemitter) here, will the implementation be more elegant?